flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java

```diff
@@ -126,15 +125,36 @@ public T getDispatcher() {
 		return webMonitorEndpoint;
 	}
 
-	@Override
-	public CompletableFuture<Void> closeAsync() {
+	/**
+	 * Deregister the Flink application from the resource management system by signalling
+	 * the {@link ResourceManager}.
+	 *
+	 * @param applicationStatus to terminate the application with
+	 * @param diagnostics additional information about the shut down, can be {@code null}
+	 * @return Future which is completed once the shut down
+	 */
+	public CompletableFuture<Void> deregisterApplicationAndClose(
+			final ApplicationStatus applicationStatus,
+			final @Nullable String diagnostics) {
+
 		if (isRunning.compareAndSet(true, false)) {
-			return FutureUtils.composeAfterwards(webMonitorEndpoint.closeAsync(), this::closeAsyncInternal);
+			final CompletableFuture<Void> closeWebMonitorAndRegisterAppFuture =
+				FutureUtils.composeAfterwards(webMonitorEndpoint.closeAsync(), () -> deregisterApplication(applicationStatus, diagnostics));
```

Comment:
are you intentionally changing the shutdown order? (Previous: webmonitor -> this::closeAsync -> deregister; now: webmonitor -> deregister -> this::closeAsync)

Reply:
Yes, because deregister is an RPC. If you shutdown the ResourceManager before deregistration, the message is lost.